src/Audience/Primitives.php

```diff
@@ -0,0 +1,244 @@
+                    return 1 === preg_match($b, $a);
+                }, $isTrace);
+
+            case '==':
+                return self::numberFun($args[0], $args[1], static function ($a, $b) {
+                    return $a === $b;
+                }, $isTrace);
+
+            case '!=':
+                return self::numberFun($args[0], $args[1], static function ($a, $b) {
+                    return $a !== $b;
+                }, $isTrace);
+
+            case '<':
+                return self::numberFun($args[0], $args[1], static function ($a, $b) {
+                    return $a < $b;
+                }, $isTrace);
+
+            case '<=':
+                return self::numberFun($args[0], $args[1], static function ($a, $b) {
+                    return $a <= $b;
+                }, $isTrace);
+
+            case '>':
+                return self::numberFun($args[0], $args[1], static function ($a, $b) {
+                    return $a > $b;
+                }, $isTrace);
+
+            case '>=':
+                return self::numberFun($args[0], $args[1], static function ($a, $b) {
+                    return $a >= $b;
+                }, $isTrace);
+
+            case 'number-attribute':
+                return self::getInEnvNumber($args[0], $environment, $isTrace);
+
+            case 'string-attribute':
+                return self::getInEnvString($args[0], $environment, $isTrace);
+
+            case 'bool-attribute':
+                return self::getInEnvBool($args[0], $environment, $isTrace);
+
+            default:
+                return self::isError(sprintf('Primitive %s is not an implemented primitive.', $primitive), $isTrace);
+        }
+    }
+
+    /**
+     * @param mixed $a
+     * @param mixed $b
+     * @return bool|array<mixed>
+     * @throws \Exception
+     */
+    private static function stringFun($a, $b, callable $function, bool $isTrace = false)
+    {
+        if (!is_string($a) || !is_string($b))
+
+            return self::isError('expected string arguments', $isTrace);
+
+        return $function($a, $b);
+    }
+
+    /**
+     * @param mixed $a
+     * @param mixed $b
+     * @return bool|array<mixed>
+     * @throws \Exception
+     */
+    private static function numberFun($a, $b, callable $function, bool $isTrace = false)
+    {
+        if (!is_numeric($a) || !is_numeric($b))
+
+            return self::isError('expected number arguments', $isTrace);
+
+        return $function((float)$a, (float)$b);
+    }
+
+    /**
+     * @param mixed $name
+     * @param array<mixed> $environment
+     * @return numeric|array<string,string>
+     * @throws \Exception
+     */
+    private static function getInEnvNumber($name, array $environment, bool $isTrace = false)
+    {
+        if (!is_string($name)) {
+            return self::isError('can only look up string names', $isTrace);
+        }
+
+        if (!isset($environment[$name]) || !is_numeric($environment[$name])) {
```

Comment:
NOTE: this review comment is not too important, just adding a note -- in the Node.js version we let the type `Environment` contains a property `attributes` in which to look up these custom attributes. In this PHP code, we get the value directly in the "environment", not inside environment["attributes"]. The reason I built it like that in the TypeScript code was to "namespace" the environment if we add to it later, to avoid collisions with custom attributes from client code. But I don't think that was necessary, should we decide to expand the environment inside here, we can just ensure the Client API maintains backwards compatibility for providing custom attributes. If we decide later that we want to namespace the environment, we can just wrap it and handle that fine.